thriftclient/client_pool.go

```diff
@@ -0,0 +1,211 @@
+package thriftclient
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/apache/thrift/lib/go/thrift"
+	"github.com/go-kit/kit/metrics"
+
+	"github.com/reddit/baseplate.go/clientpool"
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/metricsbp"
+)
+
+// ClientPoolConfig is the configuration struct for creating a new ClientPool.
+type ClientPoolConfig struct {
+	// ServiceSlug is a short identifier for the thrift service you are creating
+	// clients for.  The preferred convention is to take the service's name,
+	// remove the 'Service' prefix, if present, and convert from camel case to
+	// all lower case, hyphen separated.
+	//
+	// Example:
+	//		AuthenticationService -> authentication
+	//		ImageUploadService -> image-upload
+	ServiceSlug string
+
+	// Addr is the address of a thrift service.  Addr must be in the format
+	// ${host}:${port}
+	Addr string
+
+	// MinConnections is the minimum number of thrift connections the client
+	// pool should maintain.
+	MinConnections int
```

Comment:
Also the comment here is inaccurate because our implementation :badpokerface:

The only place min connection is used is on pool creation that we pre-populate that many clients and put into the pool. It's never used again. There's no trivial atomical way to guarantee that we always maintain the min connections in the pool and it's not really worth it to do it the non-trivial way.

Reply:
Maybe I should call it `NumInitialConnection`?

Reply:
works for me. we can also rename it in the `clientpool` interface :)

Reply:
Will do